src/async.rs

```diff
@@ -40,16 +40,20 @@ impl Signal for AsyncSignal {
 }
 
 impl<T> Hook<T, AsyncSignal> {
-    fn update_waker(&self, cx_waker: &Waker) {
+    // Update the hook to point to the given Waker.
+    // Returns whether the hook has been previously awakened
+    fn update_waker(&self, cx_waker: &Waker) -> bool {
```

Comment:
I have a question. Does the lock guard obtained in `self.1.waker.lock().will_wake()` last only in the if statement condition evaluation? Or it last for the whole if-statement including its body?

If it is the latter, I found the code confusing, because we are locking again at Line-48 (`*self.1.waker.lock() = cx_waker.clone()`).

If it is the former, then I am not sure if this is correct in the following race:

    Time  Thread-1 (calling update_waker)             Thread-2 (calling fire)
    t0    execute Line-46 and `woken` is false
    t1    the if at Line-47 evaluate to true
    t2                                                call fire executing Line 33 to 35
    t3    execute Line-48
    t4    execute Line-52

Then, we ended up not calling `cx_waker.wake_by_ref()`. That is at least contradicting with what the comment is trying to avoid.

I also have a question about the comment. Are we really swapping the wakers? What I see is that we just overwrite `self.1.waker` with the new waker, there is no swapping `cx_waker` to the old waker that was previously held in `self.1.waker`.

Reply:
@zesterer, I think @vmingchen makes some good points here. I was trying to avoid adding another atomic load here, but it looks like it might be unavoidable to prevent the race described above. 

Could you advise?

Reply:
Thanks @Restioson! I went ahead and implemented your solution that locks once and holds the guard until the function returns. I did not change the `wake` function since I'm not sure I understand the specifics.